pkg/apps/application-list.jsx

```diff
@@ -0,0 +1,168 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+
+import PackageKit from "./packagekit";
+import { left_click, icon_url, show_error, launch, ProgressBar, CancelButton } from  "./utils.jsx";
+
+const _ = cockpit.gettext;
+
+class ApplicationRow extends React.Component {
+    constructor() {
+        super();
+        this.state = { progress: false };
+    }
+
+    render() {
+        var self = this;
+        var comp = self.props.comp;
+        var state = self.state;
+
+        function navigate() {
+            cockpit.location.go(comp.id);
+        }
+
+        function action(func, arg, progress_title) {
+            self.setState({ progress_title: progress_title });
+            func(arg, (data) => self.setState({ progress: data })).
+                always(() => self.setState({ progress: false })).
+                fail(show_error);
+        }
+
+        function install() {
+            action(PackageKit.install, comp.pkgname, _("Installing"));
+        }
+
+        function remove() {
+            action(PackageKit.remove, comp.file, _("Removing"));
+        }
+
+        var name, summary_or_progress, button;
+
+        if (comp.installed) {
+            name = <a onClick={left_click(() => launch(comp))}>{comp.name}</a>;
+        } else {
+            name = comp.name;
+        }
+
+        if (state.progress) {
+            summary_or_progress = <ProgressBar title={state.progress_title} data={state.progress}/>;
+            button = <CancelButton data={state.progress}/>;
+        } else {
+            if (state.error) {
+                summary_or_progress = (
+                    <div>
+                        {comp.summary}
+                        <div className="alert alert-danger alert-dismissable">
+                            <button className="close" aria-hidden="true"
+                                    onClick={left_click(() => { this.setState({ error: null }) })}>
+                                <span className="pficon pficon-close"/>
+                            </button>
+                            <span className="pficon pficon-error-circle-o"/>
+                            {state.error}
+                        </div>
+                    </div>
+                );
+            } else {
+                summary_or_progress = comp.summary;
+            }
+
+            if (comp.installed) {
+                button = <button className="btn btn-danger" onClick={left_click(remove)}>{_("Remove")}</button>;
+            } else {
+                button = <button className="btn btn-default" onClick={left_click(install)}>{_("Install")}</button>;
+            }
```

Comment:
Could this be folded into the `if (comp.installed)` from above? it's a bit confusing to have several "parallel" if/then blocks.

Reply:
Actually, this can overwrite `button` values from the [`if (state.progress) {`](https://github.com/cockpit-project/cockpit/pull/7076/files#diff-9ec848966c668ade3ecb2918734c94a8R66) section. The flow could be a bit clearer indeed.

Reply:
Hmm, this isn't parallel with the previous `if (comp.installed)`, it is in the `no state.progress` branch.  The `button` is set exactly once.

If we move it up, we would have to duplicate the `if (state.progress)` test.  That doesn't seem any cleaner to me.